src/appleseed.studio/mainwindow/pythonconsole/linenumberarea.cpp

```diff
@@ -0,0 +1,109 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2017 Gleb Mishchenko, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "linenumberarea.h"
+
+// appleseed.studio headers.
+#include "pythoneditor.h"
+
+// Qt headers.
+#include <QPainter>
+#include <QTextBlock>
+
+
+namespace appleseed {
+namespace studio {
+
+LineNumberArea::LineNumberArea(PythonEditor* parent)
+  : QWidget(parent), editor(parent)
+{
+    connect(editor, SIGNAL(blockCountChanged(int)), this, SLOT(slot_update_area_width()));
+    connect(editor, SIGNAL(updateRequest(QRect, int)), this, SLOT(slot_update_area(QRect, int)));
+
+    slot_update_area_width();
+};
+
+void LineNumberArea::slot_update_area_width()
+{
+    setFixedWidth(area_width());
+    editor->set_left_margin(area_width());
+}
+
+void LineNumberArea::slot_update_area(const QRect& rect, int dy)
+{
+    if (dy)
+        scroll(0, dy);
+    else
+        update(0, rect.y(), area_width(), rect.height());
+
+    if (rect.contains(editor->viewport()->rect()))
+        slot_update_area_width();
+}
+
+int LineNumberArea::area_width()
+{
+    int digits = 1;
+    int max = qMax(1, editor->blockCount());
+    while(max >= 10) {
+        max /= 10;
+        ++digits;
+    }
+
+    return 3 + fontMetrics().width(QLatin1Char('9')) * digits;
+}
+
+void LineNumberArea::paintEvent(QPaintEvent* event)
+{
+    QPainter painter(this);
+    painter.fillRect(event->rect(), QColor(53, 53, 53));
```

Comment:
I guess we can live with hardcoded values for now, it would be nice eventually to investigate how we could define "standalone" colors in the stylesheets and use them from the C++ code.